Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp

```diff
@@ -177,6 +177,199 @@ bool TerrainSystem::InWorldBounds(float x, float y) const
+    // position that falls in the same area as the previous positions. This consumes lesser memory
+    // than sorting the points into separate lists and handling putting them back together.
+    // This may be sub optimal if the points are randomly distributed in the list as opposed
+    // to points in the same area id being close to each other.
+    size_t windowStart = 0;
+    size_t windowEnd = 0;
+    const size_t numPositions = inPositions.size();
+    for(int i = 1; i < numPositions; i++)
+    {
+        AZ::EntityId areaId = FindBestAreaEntityAtPosition(inPositions[i].GetX(), inPositions[i].GetY(), bounds);
+        bool queryHeights = false;
+        if (areaId == prevAreaId)
+        {
+            // Update window end to current position.
+            // If it's the last position, submit the query.
+            windowEnd = i;
+            if (windowEnd == numPositions - 1)
+            {
+                queryHeights = true;
+            }
+        }
+        else
+        {
+            queryHeights = true;
+        }
+
+        if (queryHeights)
+        {
+            // If the area id is a default entity id, it usually means the
+            // position is outside world bounds.
+            if (prevAreaId != AZ::EntityId())
+            {
+                size_t spanLength = (windowEnd - windowStart) + 1;
+                queryCallback(AZStd::span<AZ::Vector3>(inPositions.begin() + windowStart, spanLength),
+                    AZStd::span<AZ::Vector3>(outPositions.begin() + windowStart, spanLength),
+                    AZStd::span<bool>(outTerrainExists.begin() + windowStart, spanLength),
+                    AZStd::span<AzFramework::SurfaceData::SurfaceTagWeightList>(outSurfaceWeights.begin() + windowStart, spanLength),
+                    prevAreaId);
+            }
+
+            // Reset the window to start at the current position. Set the new area
+            // id on which to run the next query.
+            windowStart = windowEnd = i;
+            prevAreaId = areaId;
+        }   
+    }
+}
+
+void TerrainSystem::GetHeightsSynchronous(const AZStd::span<AZ::Vector3>& inPositions, Sampler sampler, 
+    AZStd::span<float> heights, AZStd::span<bool> terrainExists) const
+{
+    AZStd::shared_lock<AZStd::shared_mutex> lock(m_areaMutex);
+
+    AZStd::vector<AZ::Vector3> outPositions;
+    AZStd::vector<bool> outTerrainExists;
+
+    size_t indexStepSize = 1;
+    if (sampler == AzFramework::Terrain::TerrainDataRequests::Sampler::BILINEAR)
+    {
+        // outPositions holds the iterators to results of the bulk queries.
+        // In the case of the bilinear sampler, we'll be making 4 queries per
+        // input position.
+        outPositions.resize(inPositions.size() * 4);
+        outTerrainExists.resize(inPositions.size() * 4);
+        indexStepSize = 4;
+    }
+    else
+    {
+        outPositions.resize(inPositions.size());
+        outTerrainExists.resize(inPositions.size());
+    }
```

Comment:
As mentioned above, these would become "reserve()" for outPositions instead of "resize()" if you change the methods that fill them out to use emplace_back(), which would then avoid constructing a bunch of default Vector3s that will get overwritten.

Reply:
Done